### PR TITLE
Replace EVP_PKEY_supports_digest_nid

### LINKED ISSUE #14343: EVP_PKEY_supports_digest_nid does not work with provider-side keys
I am facing a regression of #7348. During TLS negotiation is the TPM provider asked to sign using an unsupported hash, provided by the default provider. The TPM cannot sign using external hash algorithms, so despite a hash algorithm is implemented by some other provider it cannot be used for digest-signing using TPM-based keys.

To solve this in a generic way, the `EVP_PKEY_supports_digest_nid` has been created some years ago and used in [`ssl/t1_lib.c:check_cert_usable`](https://github.com/openssl/openssl/blob/master/ssl/t1_lib.c#L3045) to check whether a particular hash is compatible with a particular sigalg. It is however based on `evp_pkey_asn1_ctrl`, which works for legacy keys only, so the old problem re-appears after migration to providers.

I found few possibilities how this could be solved "in the provider world", but none of them is perfect. The trouble is there seems to be no mechanism to "check" something.
 * The `EVP_PKEY_CTX_set_signature_md` could be (mis)used to check compatibility of a hash, but this requires creation of a temporary PKEY_CTX and initialization of a sign/verify operation (e.g. `EVP_PKEY_sign_init`). These steps can be seen as a waste of computing resources, decreasing the performance as it needs to be called for each candidate sigalg.
 * There could be a new KEYMGMT settable parameter such as `OSSL_PKEY_PARAM_MD_SUPPORTED` that will only test validity of the argument, instead of setting it. But this is also a slight misuse of the API.
 * Or there could be a completely new KEYMGMT function to check various assertions (`OSSL_FUNC_KEYMGMT_ASSERT_PARAMS`) that would consume e.g. the `OSSL_PKEY_PARAM_MD_SUPPORTED`, verify the assertion ("this MD is supported") and return 0/1 accordingly. It seems to me that e.g. `ASN1_PKEY_CTRL_CMS_IS_RI_TYPE_SUPPORTED` might be also unsupported by providers, so this might be usable also for others.

The `OSSL_FUNC_KEYMGMT_ASSERT_PARAMS` function feels to be the cleanest solution, but is it acceptable to introduce yet another KEYMGMT function? (It's not very invasive and it's quite generic hence reusable.) Or do you see other options how the `EVP_PKEY_supports_digest_nid` can be implemented for non-legacy?

### PR DESCRIPTION
**Pull request by t8m**
The EVP_PKEY_supports_digest_nid() is renamed to EVP_PKEY_digestsign_supports_digest() and implemented via EVP_DigestSignInit_ex().

Fixes #14343

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by t8m** [comment]
IMO there is a reason. Other applications might have exactly the same need to identify digests that are acceptable to be used with a given key.

The API doesn't preclude a different implementation for example by having an extra provider API call.

7. 5. 2021 21:27, 21:27, Richard Levitte ***@***.***> napsal/a:

File: doc/man3/EVP_PKEY_digestsign_supports_digest.pod [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+L<EVP_DigestSignInit_ex(3)>,
+
+=head1 HISTORY
+
+The EVP_PKEY_digestsign_supports_digest() function was added in OpenSSL 3.0.
+
+=head1 COPYRIGHT
+
+Copyright 2006-2021 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (slontis) [inline doc/man3/EVP_PKEY_digestsign_supports_digest.pod:37]: Just 2021

File: crypto/evp/p_lib.c [hunk 2]
```diff
@@ -1335,23 +1335,21 @@ int EVP_PKEY_get_group_name(const EVP_PKEY *pkey, char *gname, size_t gname_sz,
...
-            return (nid == default_nid);
-        /* zero is an error from EVP_PKEY_get_default_digest_nid() */
-        if (rv == 0)
-            return -1;
-    }
+    if ((ctx = EVP_MD_CTX_new()) == NULL)
+        return -1;
+
+    ERR_set_mark();
+    rv = EVP_DigestSignInit_ex(ctx, NULL, name, libctx,
```
Comment (slontis) [inline crypto/evp/p_lib.c:1348]: Is this equivalent to what it did before?
ASN1_PKEY_CTRL_DEFAULT_MD_NID for ec for example returned SM3 or SHA256..
rsa pss returned md?

Comment (levitte) [reply crypto/evp/p_lib.c:1348]: We will have to decide to support, or not, signatures that combine those keytypes with those digests.
I *assume* that if `ASN1_PKEY_CTRL_DEFAULT_MD_NID` returned a certain NID, `EVP_DigestSignIni` using a digest produced with that NID wouldn't refuse it.

So, if it's not equivalent, we either need to restore support for the "offending" digest, or document the breaking change.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master fixing the trivial libcrypto.num conflict during the merge. Thank you for the reviews.

**Pull request closed without merging**